src/main/java/uk/gov/register/serialization/CommandParser.java

```diff
@@ -45,40 +60,75 @@ public RegisterCommand newCommand(String s){
+                    } catch (JsonParseException jpe) {
                         LOG.error("failed to parse json: " + parts[1]);
                         throw new SerializedRegisterParseException("failed to parse json: " + parts[1], jpe);
                     } catch (IOException e) {
-                        LOG.error("",e);
+                        LOG.error("", e);
                         throw new UncheckedIOException(e);
                     }
                 } else {
                     LOG.error("add item line must have 2 elements, was: " + s);
                     throw new SerializedRegisterParseException("add item line must have 2 elements, was: " + s);
                 }
+                break;
             case "append-entry":
                 if (parts.length == 3) {
                     Entry entry = new Entry(0, stripPrefix(parts[2]), Instant.parse(parts[1]));
-                    return new AppendEntryCommand(entry);
+                    entries.put(position, entry);
+                    updateEntryHashes(entry.getSha256hex(), position++);
                 } else {
                     LOG.error("append entry line must have 3 elements, was : " + s);
                     throw new SerializedRegisterParseException("append entry line must have 3 elements, was : " + s);
                 }
+                break;
             case "assert-root-hash":
                 if (parts.length == 2) {
-                    return new AssertRootHashCommand(new RegisterProof(parts[1]));
+                    RegisterProof registerProof = new RegisterProof(parts[1]);
+                    proofs.put(position++, registerProof);
                 } else {
                     LOG.error("assert root hash line must have 1 elements, was : " + s);
                     throw new SerializedRegisterParseException("assert root hash line must have 1 elements, was : " + s);
                 }
+                break;
             default:
                 LOG.error("line must begin with legal command not:" + commandName);
                 throw new SerializedRegisterParseException("line must begin with legal command not: " + commandName);
         }
     }
 
-    private  String stripPrefix(String hashField) {
+    public Iterator<RegisterCommand> getCommands() {
+
+        validateOrphanItems();
+
+        return IntStream.range(0, position).mapToObj(i -> {
+            if (entries.containsKey(i)) {
+                return new AppendEntryCommand(entries.get(i));
+            } else if (items.containsKey(i)) {
+                return new AddItemCommand(items.get(i));
+            } else if (proofs.containsKey(i)) {
+                return new AssertRootHashCommand(proofs.get(i));
+            } else {
+                throw new RuntimeException("No command found for position " + String.valueOf(i));
+            }
+        }).iterator();
+    }
+
+    private void validateOrphanItems() {
+        Set<Item> orphanItems = items.entrySet().stream().filter(kv -> {
```

Comment:
Would it be possible just to use a `HashMap` or similar that stores item hashes against number of times that item hash has been referenced by `append-entry` commands? Then you can keep track of what is and isn't being used, and also enforces the required ordering.